semgrep/semgrep/join_rule.py

```diff
@@ -0,0 +1,404 @@
+
+
+def create_model_map(
+    semgrep_results: List[Dict[str, Any]]
+) -> Dict[str, Type[BaseModel]]:
+    collections: Dict[str, List[Dict]] = group(
+        semgrep_results, key=lambda item: item.get("check_id")
+    )
+    model_map: Dict[str, Type[BaseModel]] = {}
+    for name, findings in collections.items():
+        metavars = set()
+        for finding in findings:
+            metavars.update(finding.get("extra", {}).get("metavars", {}).keys())
+        model_fields = ["path"] + list(metavars)
+        model_class = model_factory(camel_case(name), model_fields)
+        model_map[name] = model_class
+    return model_map
+
+
+def populate_data(
+    semgrep_results: List[Dict[str, Any]], model_map: Dict[str, Type[BaseModel]]
+) -> None:
+    collections = group(semgrep_results, key=lambda item: item.get("check_id"))
+    for name, findings in collections.items():
+        for finding in findings:
+            model_map[name].create(
+                path=finding.get("path"),
+                raw=json.dumps(finding),
+                **{
+                    metavar: content.get("abstract_content").strip().strip("\"'")
+                    for metavar, content in finding.get("extra", {})
+                    .get("metavars", {})
+                    .items()
+                },
+            )
+
+
+def main(
+    join_rule: Dict[str, Any],
+    targets: List[Path],
+) -> Tuple[List[RuleMatch], List[SemgrepError]]:
+    join_contents = join_rule.get("join", {})
+    semgrep_config_strings = [ref.get("rule") for ref in join_contents.get("refs", [])]
+    config_map = create_config_map(semgrep_config_strings)
+
+    join_rule_refs: List[Ref] = [
+        Ref(
+            id=config_map[ref.get("rule")].id,
+            renames={
+                rename.get("from"): rename.get("to")
+                for rename in ref.get("renames", [])
+            },
+            alias=ref.get("as"),
+        )
+        for ref in join_contents.get("refs", [])
+    ]
+    refs_lookup = {ref.id: ref for ref in join_rule_refs}
+    alias_lookup = {ref.alias: ref.id for ref in join_rule_refs}
+
+    try:
+        conditions = [
+            Condition.parse(condition_string)
+            for condition_string in join_contents.get("on", [])
+        ]
+    except InvalidConditionError as e:
+        return [], [e]
+
+    # Run Semgrep
+    with tempfile.NamedTemporaryFile() as rule_path:
+        yaml.dump({"rules": [rule.raw for rule in config_map.values()]}, rule_path)
+        rule_path.flush()
+        output = semgrep.semgrep_main.invoke_semgrep(
+            config=Path(rule_path.name),
+            targets=targets,
+            no_rewrite_rule_ids=True,
+            optimizations="all",
+        )
+
+    assert isinstance(output, dict)  # placate mypy
+
+    results = output.get("results", [])
+    errors = output.get("errors", [])
+
+    parsed_errors = []
+    for error_dict in errors:
+        try:
+            del error_dict["code"]
```

Comment:
why do these keys need to be deleted?

Reply:
This was a hack to reconstitute errors after they've been JSONified as output. `SemgrepError` subclasses define the `level` and `code` as class properties, which means they aren't accepted as arguments when instantiated. They'll still be properly populated because it's using the class properties according to the instantiated class.

I'll add a comment explaining this.

I think a more proper implementation to avoid having to reconstitute errors would be to return the actual result and error objects from the Semgrep API instead of JSON. Or add a flag indicating which is preferred.  But in the meantime, I figured this was... well, it works. 😅